File: Server/PluginSDK/Go/tests/sdk_test.go
package tests

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	pluginsdk "sprakcapture/plugin-sdk-go"
	"sprakcapture/plugin-sdk-go/bridgegen"
)

var registerTestPluginOnce sync.Once
var stoppedTestRuntimes atomic.Int32
var destroyedTestRuntimes atomic.Int32

// registerTestPlugin 注册测试进程唯一插件；处理器按 payload.delayMs 延迟以验证并发乱序与 stop 屏障。
func registerTestPlugin() {
	registerTestPluginOnce.Do(func() {
		pluginsdk.Register(func(initContext pluginsdk.InitContext) (pluginsdk.Plugin, error) {
			var lifecycle struct {
				PanicStop bool `json:"panicStop"`
			}
			if err := json.Unmarshal(initContext.Configuration, &lifecycle); err != nil {
				return pluginsdk.Plugin{}, err
			}
			return pluginsdk.Plugin{Handle: func(ctx context.Context, invocation pluginsdk.Invocation) (pluginsdk.Action, error) {
				if err := ctx.Err(); err != nil {
					return pluginsdk.Action{}, err
				}
				var timing struct {
					DelayMS int `json:"delayMs"`
				}
				if err := json.Unmarshal(invocation.Envelope.Payload, &timing); err != nil {
					return pluginsdk.Action{}, err
				}
				if timing.DelayMS > 0 {
					timer := time.NewTimer(time.Duration(timing.DelayMS) * time.Millisecond)
					defer timer.Stop()
					select {
					case <-timer.C:
					case <-ctx.Done():
						return pluginsdk.Action{}, ctx.Err()
					}
				}
				event, err := pluginsdk.ParseBinaryEvent(invocation)
				if err != nil {
					return pluginsdk.Action{}, err
				}
				return pluginsdk.ModifyBytes(invocation, bytes.ToUpper(event.Bytes))
			}, OnStop: func() {
				stoppedTestRuntimes.Add(1)
				if lifecycle.PanicStop {
					panic("停止夹具 panic")
				}
			}, OnDestroy: func() {
				destroyedTestRuntimes.Add(1)
			}}, nil
		})
	})
}

// TestExpiredEnvelopeDeadlineIsInformational 验证过期 deadline 仅供参考，不会触发 SDK 强制取消。
func TestExpiredEnvelopeDeadlineIsInformational(t *testing.T) {
	registerTestPlugin()
	invocation := workerInvocation("event-expired", []byte("active"), 0)
	invocation["envelope"].(map[string]any)["deadlineUnixMs"] = uint64(1)
	input := encodeLines(t,
		map[string]any{"type": "initialize", "apiVersion": 2, "manifest": map[string]any{}, "configuration": map[string]any{}},
		map[string]any{"type": "invoke", "requestId": uint64(9), "invocation": invocation},
		map[string]any{"type": "stop"},
	)
	var output bytes.Buffer
	if err := pluginsdk.RunJSONL(context.Background(), strings.NewReader(input), &output); err != nil {
		t.Fatalf("过期参考 deadline 不应取消调用：%v", err)
	}
	if !strings.Contains(output.String(), `"type":"result"`) || !strings.Contains(output.String(), `"requestId":9`) {
		t.Fatalf("过期参考 deadline 未返回结果：%s", output.String())
	}
}

// TestWorkerRequestIDDomain 验证零值与最大安全整数无损回传，超出跨语言安全域的值明确失败。
func TestWorkerRequestIDDomain(t *testing.T) {
	registerTestPlugin()
	const maximumSafeRequestID uint64 = 9_007_199_254_740_991
	testCases := []struct {
		name      string
		requestID uint64
		rejected  bool
	}{
		{name: "零值", requestID: 0},
		{name: "最大安全整数", requestID: maximumSafeRequestID},
		{name: "超出安全整数", requestID: maximumSafeRequestID + 1, rejected: true},
	}
	for _, testCase := range testCases {
		t.Run(testCase.name, func(t *testing.T) {
			invocation := workerInvocation("event-request-id", []byte("id"), 0)
			input := encodeLines(t,
				map[string]any{"type": "initialize", "apiVersion": 2, "manifest": map[string]any{}, "configuration": map[string]any{}},
				map[string]any{"type": "invoke", "requestId": testCase.requestID, "invocation": invocation},
				map[string]any{"type": "stop"},
			)
			var output bytes.Buffer
			err := pluginsdk.RunJSONL(context.Background(), strings.NewReader(input), &output)
			if testCase.rejected {
				if err == nil || !strings.Contains(err.Error(), "安全整数") {
					t.Fatalf("超域 requestId 应明确失败：%v", err)
				}
				return
			}
			if err != nil {
				t.Fatalf("安全 requestId 不应失败：%v", err)
			}
			expected := fmt.Sprintf(`"requestId":%d`, testCase.requestID)
			if !strings.Contains(output.String(), expected) {
				t.Fatalf("requestId 未按 JSON 数字无损回传：%s", output.String())
			}
		})
	}
}

// TestWorkerStopPanicReturnsError 验证 worker stop 回调异常被转换为错误而非越过运行时边界。
func TestWorkerStopPanicReturnsError(t *testing.T) {
	registerTestPlugin()
	input := encodeLines(t,
		map[string]any{"type": "initialize", "apiVersion": 2, "manifest": map[string]any{}, "configuration": map[string]any{"panicStop": true}},
		map[string]any{"type": "stop"},
	)
	var output bytes.Buffer
	err := pluginsdk.RunJSONL(context.Background(), strings.NewReader(input), &output)
	if err == nil || !strings.Contains(err.Error(), "OnStop") {
		t.Fatalf("stop panic 应返回明确错误：%v", err)
	}
}

// TestPacketHelpers 验证分包、合包与跨 chunk 长度前缀解析保持全部字节和顺序。
func TestPacketHelpers(t *testing.T) {
	packets, err := pluginsdk.SplitPackets([]byte("abcdefgh"), 3)
	if err != nil {
		t.Fatalf("切分负载：%v", err)
	}
	joined, err := pluginsdk.JoinPackets(packets)
	if err != nil || string(joined) != "abcdefgh" {
		t.Fatalf("合并负载错误：%q, %v", joined, err)
	}
	framer, err := pluginsdk.NewLengthPrefixedFrames(32)
	if err != nil {
		t.Fatalf("创建分帧器：%v", err)
	}
	encoded, err := framer.Encode([]byte("hello"))
	if err != nil {
		t.Fatalf("重封包：%v", err)
	}
	first, err := framer.Push(encoded[:2])
	if err != nil || len(first) != 0 {
		t.Fatalf("半前缀不得产生帧：%v, %v", first, err)
	}
	second, err := framer.Push(encoded[2:])
	if err != nil || len(second) != 1 || string(second[0]) != "hello" {
		t.Fatalf("跨 chunk 解析错误：%v, %v", second, err)
	}
}

// TestJSONLWorker 验证 initialize→ready、invoke→result、stop 的完整生产协议与字节改写。
func TestJSONLWorker(t *testing.T) {
	registerTestPlugin()
	invocation := map[string]any{
		"pluginId":   "example.go-worker",
		"moduleId":   "rewrite",
		"moduleKind": "streamTransformer",
		"envelope": map[string]any{
			"apiVersion": "2.0.0", "eventId": "event-go-1", "stage": "udpDatagram",
			"serviceGeneration": 1, "recordingGeneration": 1,
			"pluginInstanceId": "example.go-worker@1.0.0#1", "deadlineUnixMs": uint64(4102444800000),
			"context": map[string]any{"transport": "udp", "interceptionMode": "intercept", "labels": []string{}},
			"payload": map[string]any{"bytes": pluginsdk.ByteArray("hello"), "endOfStream": false},
		},
	}
	input := encodeLines(t,
		map[string]any{"type": "initialize", "apiVersion": 2, "manifest": map[string]any{}, "configuration": map[string]any{}},
		map[string]any{"type": "invoke", "requestId": uint64(1), "invocation": invocation},
		map[string]any{"type": "stop"},
	)
	var output bytes.Buffer
	if err := pluginsdk.RunJSONL(context.Background(), strings.NewReader(input), &output); err != nil {
		t.Fatalf("运行 JSONL worker：%v", err)
	}
	decoder := json.NewDecoder(&output)
	var ready map[string]any
	var result map[string]any
	if err := decoder.Decode(&ready); err != nil || ready["type"] != "ready" || ready["apiVersion"] != float64(2) {
		t.Fatalf("ready 响应错误：%v, %v", ready, err)
	}
	if err := decoder.Decode(&result); err != nil || result["type"] != "result" || result["requestId"] != float64(1) {
		t.Fatalf("result 响应错误：%v, %v", result, err)
	}
	action := result["action"].(map[string]any)
	patch := action["patch"].([]any)
	bytesValue := patch[0].(map[string]any)["value"].(map[string]any)["bytes"].([]any)
	if len(bytesValue) != 5 || bytesValue[0] != float64('H') || bytesValue[4] != float64('O') {
		t.Fatalf("改写动作字节错误：%v", bytesValue)
	}
}

// TestConcurrentJSONLWorker 验证显式并发允许按完成顺序返回，并在 stop 前等待慢调用完整写出。
func TestConcurrentJSONLWorker(t *testing.T) {
	registerTestPlugin()
	stoppedBefore := stoppedTestRuntimes.Load()
	destroyedBefore := destroyedTestRuntimes.Load()
	slow := workerInvocation("event-slow", []byte("slow"), 80)
	fast := workerInvocation("event-fast", []byte("fast"), 0)
	input := encodeLines(t,
		map[string]any{"type": "initialize", "apiVersion": 2, "manifest": map[string]any{}, "configuration": map[string]any{}},
		map[string]any{"type": "invoke", "requestId": uint64(1), "invocation": slow},
		map[string]any{"type": "invoke", "requestId": uint64(2), "invocation": fast},
		map[string]any{"type": "stop"},
	)
	var output bytes.Buffer
	if err := pluginsdk.RunJSONLWithOptions(
		context.Background(),
		strings.NewReader(input),
		&output,
		pluginsdk.WorkerOptions{MaxConcurrentInvocations: 2},
	); err != nil {
		t.Fatalf("运行并发 JSONL worker：%v", err)
	}
	decoder := json.NewDecoder(&output)
	responses := make([]map[string]any, 0, 3)
	for len(responses) < 3 {
		var response map[string]any
		if err := decoder.Decode(&response); err != nil {
			t.Fatalf("解析并发响应：%v", err)
		}
		responses = append(responses, response)
	}
	var unexpected map[string]any
	if err := decoder.Decode(&unexpected); !errors.Is(err, io.EOF) {
		t.Fatalf("并发 worker 输出了额外响应：%v, %v", unexpected, err)
	}
	if len(responses) != 3 || responses[0]["type"] != "ready" {
		t.Fatalf("并发响应数量或 ready 错误：%v", responses)
	}
	if responses[1]["requestId"] != float64(2) || responses[2]["requestId"] != float64(1) {
		t.Fatalf("并发完成顺序或 requestId 错误：%v", responses)
	}
	if stoppedTestRuntimes.Load() != stoppedBefore+1 || destroyedTestRuntimes.Load() != destroyedBefore+1 {
		t.Fatalf("stop/destroy 生命周期没有各执行一次：stop=%d, destroy=%d", stoppedTestRuntimes.Load()-stoppedBefore, destroyedTestRuntimes.Load()-destroyedBefore)
	}
}

// TestActionConstructors 验证全部稳定宿主动作的字段、输入校验及二进制字段保留语义。
func TestActionConstructors(t *testing.T) {
	invocation := pluginsdk.Invocation{Envelope: pluginsdk.EventEnvelope{
		EventID: "event-actions",
		Payload: json.RawMessage(`{"bytes":[1,2],"endOfStream":true,"channel":"client"}`),
	}}
	if action := pluginsdk.Continue(invocation); action.Action != pluginsdk.ActionContinue {
		t.Fatalf("continue 动作错误：%v", action)
	}
	if action := pluginsdk.Hold(invocation); action.Action != pluginsdk.ActionHold {
		t.Fatalf("hold 动作错误：%v", action)
	}
	if action := pluginsdk.Drop(invocation); action.Action != pluginsdk.ActionDrop {
		t.Fatalf("drop 动作错误：%v", action)
	}

	modified, err := pluginsdk.ModifyBytes(invocation, []byte{3, 4})
	if err != nil {
		t.Fatalf("构造二进制修改动作：%v", err)
	}
	var operation map[string]json.RawMessage
	if err := json.Unmarshal(modified.Patch[0], &operation); err != nil {
		t.Fatalf("解析修改补丁：%v", err)
	}
	var payload map[string]json.RawMessage
	if err := json.Unmarshal(operation["value"], &payload); err != nil {
		t.Fatalf("解析修改后的 payload：%v", err)
	}
	if string(payload["bytes"]) != "[3,4]" || string(payload["endOfStream"]) != "true" || string(payload["channel"]) != `"client"` {
		t.Fatalf("二进制修改未保留 payload 字段：%s", operation["value"])
	}

	rejected, err := pluginsdk.Reject(invocation, "不允许")
	if err != nil || rejected.Action != pluginsdk.ActionReject {
		t.Fatalf("reject 动作错误：%v, %v", rejected, err)
	}
	closed, err := pluginsdk.Close(invocation, "连接结束")
	if err != nil || closed.Action != pluginsdk.ActionClose {
		t.Fatalf("close 动作错误：%v, %v", closed, err)
	}
	if _, err := pluginsdk.Reject(invocation, "  "); err == nil {
		t.Fatal("空白 reject 原因必须失败")
	}

	annotated, err := pluginsdk.Annotate(invocation, map[string]any{"tag": "检查"})
	if err != nil || annotated.Action != pluginsdk.ActionAnnotate || len(annotated.Annotations) != 1 {
		t.Fatalf("annotate 动作错误：%v, %v", annotated, err)
	}
	redirected, err := pluginsdk.Redirect(invocation, "upstream.local", 8443)
	if err != nil || redirected.Action != pluginsdk.ActionRedirect {
		t.Fatalf("redirect 动作错误：%v, %v", redirected, err)
	}
	if _, err := pluginsdk.Redirect(invocation, "", 0); err == nil {
		t.Fatal("无效 redirect 目标必须失败")
	}
	responded, err := pluginsdk.Respond(invocation, map[string]any{"statusCode": 204})
	if err != nil || responded.Action != pluginsdk.ActionRespond {
		t.Fatalf("respond 动作错误：%v, %v", responded, err)
	}
	modifiedPayload, err := pluginsdk.ModifyPayload(invocation, map[string]any{"value": 7})
	if err != nil || modifiedPayload.Action != pluginsdk.ActionModify || len(modifiedPayload.Patch) != 1 {
		t.Fatalf("通用 modify 动作错误：%v, %v", modifiedPayload, err)
	}
}

// workerInvocation 构造带可控延迟的二进制 worker 调用，供串并行契约测试复用。
func workerInvocation(eventID string, payload []byte, delayMS int) map[string]any {
	return map[string]any{
		"pluginId": "example.go-worker", "moduleId": "rewrite", "moduleKind": "streamTransformer",
		"envelope": map[string]any{
			"apiVersion": "2.0.0", "eventId": eventID, "stage": "tcpChunk",
			"serviceGeneration": 1, "recordingGeneration": 1,
			"pluginInstanceId": "example.go-worker@1.0.0#1", "deadlineUnixMs": uint64(4102444800000),
			"context": map[string]any{"transport": "tcp", "interceptionMode": "intercept", "labels": []string{}},
			"payload": map[string]any{"bytes": pluginsdk.ByteArray(payload), "endOfStream": false, "delayMs": delayMS},
		},
	}
}

// TestBridgeGenerator 验证生成器只创建固定 ABI 桥文件并包含宿主要求的导出符号。
func TestBridgeGenerator(t *testing.T) {
	directory := t.TempDir()
	if err := bridgegen.Generate(bridgegen.Options{
		OutputDirectory: directory,
		PackageName:     "main",
		SDKImportPath:   "sprakcapture/plugin-sdk-go",
	}); err != nil {
		t.Fatalf("生成 Native 桥：%v", err)
	}
	exportBytes, err := os.ReadFile(filepath.Join(directory, "zzNativeBridge.go"))
	if err != nil {
		t.Fatalf("读取生成桥：%v", err)
	}
	if !bytes.Contains(exportBytes, []byte("//export capture_extension_init")) ||
		!bytes.Contains(exportBytes, []byte("goCaptureInvoke")) {
		t.Fatalf("生成桥缺少 ABI v2 导出")
	}
}

// encodeLines 把测试消息编码为真实 JSONL；任一序列化失败都立即终止用例。
func encodeLines(t *testing.T, messages ...any) string {
	t.Helper()
	var output bytes.Buffer
	encoder := json.NewEncoder(&output)
	for _, message := range messages {
		if err := encoder.Encode(message); err != nil {
			t.Fatalf("编码 JSONL 消息：%v", err)
		}
	}
	return output.String()
}

File: Server/PluginSDK/TS/tests/sdk.test.ts
/** 验证 TypeScript SDK 的动作、流式协议与 JSONL 运行器。 */

import assert from "node:assert/strict";
import { spawn } from "node:child_process";
import { once } from "node:events";
import { createInterface } from "node:readline";
import test from "node:test";
import { fileURLToPath } from "node:url";
import { dirname, join, resolve } from "node:path";
import { access, mkdtemp, readFile, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import { Event, Frame, LengthPrefixedCodec, ManifestBuilder, Simulator, StreamPipeline, createInvocation, definePlugin, modifyBytes, modifyPayload, redirect, reject } from "../src/index.js";

test("普通阶段函数直接构造修改动作", async () => {
  /** 注册 UDP 函数并验证 Host 期望的完整动作结构。 */
  const plugin = definePlugin();
  plugin.on("udpDatagram", (event) => modifyBytes(event, event.bytes().reverse()));
  const action = await new Simulator(plugin).invoke(createInvocation("event-1", "udpDatagram", { bytes: [1, 2, 3] }));
  assert.equal(action.action, "modify");
  assert.deepEqual(action.patch[0]?.value, { bytes: [3, 2, 1] });
});

test("终止动作拒绝纯空白参数", () => {
  /** 保持五种 SDK 的 reason/host 校验一致，避免同一作者输入产生不同动作。 */
  const event = new Event(createInvocation("event-validation", "beforeConnect", { bytes: [] }));
  assert.throws(() => reject(event, " \t\r\n"), /拒绝原因/);
  assert.throws(() => redirect(event, " \t", 443), /重定向主机/);
});

test("任意 JSON payload 保持原始类型", () => {
  /** 标量和数组可用于命令等阶段，只有 bytes 便捷函数要求对象 payload。 */
  const event = new Event(createInvocation("event-scalar", "commandInvoked", 7));
  assert.equal(event.payload, 7);
  assert.deepEqual(modifyPayload(event, ["ok"]).patch[0]?.value, ["ok"]);
  assert.throws(() => modifyBytes(event, Uint8Array.of(1)), /payload 必须是对象/);
});

test("TCP 管线保留半帧并重封包", async () => {
  /** 分两次推送一帧，确保第一块 hold 且第二块只发布完整线帧。 */
  const plugin = definePlugin();
  plugin.tcp(new StreamPipeline(() => new LengthPrefixedCodec(2)), (frame) => ({ ...frame, payload: Uint8Array.from(frame.payload, (value) => value - 32) }));
  const first = await plugin.invoke(createInvocation("event-1", "tcpChunk", { bytes: [0, 3, 97] }));
  const second = await plugin.invoke(createInvocation("event-2", "tcpChunk", { bytes: [98, 99] }));
  assert.equal(first.action, "hold");
  assert.deepEqual(second.patch[0]?.value, { bytes: [0, 3, 65, 66, 67] });
});

test("manifest 使用 sidecar 和 JavaScript 入口", () => {
  /** 确保 TypeScript 构建结果不会被错误声明为生产 Native 动态库。 */
  const manifest = new ManifestBuilder("example.ts", "示例", "1.0.0", "example").module("traffic", "trafficHandler", "udpDatagram").build("plugin.js");
  assert.deepEqual(manifest.runtime, { kind: "sidecar", entry: "plugin.js", protocolVersion: "2.0", arguments: [] });
});

test("构建输出包含可直接加载的完整示例", async () => {
  /**
   * Host 按 manifest 所在目录解析相对 entry；同时检查二者可访问，防止 tsc 只生成
   * JavaScript 却漏掉清单，导致发布包中的示例无法安装。
   */
  const currentDirectory = dirname(fileURLToPath(import.meta.url));
  const exampleDirectory = resolve(currentDirectory, "../examples/binaryProtocol");
  const manifest = JSON.parse(await readFile(join(exampleDirectory, "plugin.json"), "utf8")) as {
    runtime: { entry: string };
  };
  await access(join(exampleDirectory, manifest.runtime.entry));
  assert.equal(manifest.runtime.entry, "plugin.js");
});

test("runner 严格实现 JSONL", async () => {
  /** 启动真实 Node 子进程，验证逐行 ready/result 与 requestId 关联。 */
  const currentDirectory = dirname(fileURLToPath(import.meta.url));
  const runner = resolve(currentDirectory, "../src/runner.js");
  const example = resolve(currentDirectory, "../examples/binaryProtocol/plugin.js");
  const child = spawn(process.execPath, [runner, example], { stdio: ["pipe", "pipe", "pipe"] });
  const lines = createInterface({ input: child.stdout });
  const responses: Record<string, unknown>[] = [];
  lines.on("line", (line) => responses.push(JSON.parse(line) as Record<string, unknown>));
  child.stdin.write(`${JSON.stringify({ type: "initialize", apiVersion: 2, manifest: {}, configuration: {} })}\n`);
  child.stdin.write(`${JSON.stringify({ type: "invoke", requestId: 1, invocation: createInvocation("event-1", "tcpChunk", { bytes: [0, 1, 97] }) })}\n`);
  child.stdin.end(`${JSON.stringify({ type: "stop" })}\n`);
  const [exitCode] = await once(child, "exit");
  assert.equal(exitCode, 0);
  assert.deepEqual(responses[0], { type: "ready", apiVersion: 2 });
  assert.equal(responses[1]?.type, "result");
  assert.equal(responses[1]?.requestId, 1);
  assert.equal(typeof responses[1]?.requestId, "number");
  const result = responses[1] as { action: { patch: { value: unknown }[] } };
  assert.deepEqual(result.action.patch[0]?.value, { bytes: [0, 1, 65] });
});

test("显式并发允许乱序结果并在 stop 后释放一次", async () => {
  /** 真实子进程发送慢、快两个请求，验证原子帧、requestId 和停止等待。 */
  const currentDirectory = dirname(fileURLToPath(import.meta.url));
  const runner = resolve(currentDirectory, "../src/runner.js");
  const fixture = resolve(currentDirectory, "concurrentPlugin.js");
  const temporaryDirectory = await mkdtemp(join(tmpdir(), "capture-ts-sdk-"));
  const probePath = join(temporaryDirectory, "stopped.txt");
  try {
    const child = spawn(process.execPath, [runner, fixture, "--concurrent"], { stdio: ["pipe", "pipe", "pipe"], env: { ...process.env, PLUGIN_STOP_PROBE: probePath } });
    const lines = createInterface({ input: child.stdout });
    const responses: Record<string, unknown>[] = [];
    lines.on("line", (line) => responses.push(JSON.parse(line) as Record<string, unknown>));
    const messages = [
      { type: "initialize", apiVersion: 2, manifest: {}, configuration: {} },
      { type: "invoke", requestId: 1, invocation: createInvocation("slow-event", "udpDatagram", { bytes: [1], delayMs: 150 }) },
      { type: "invoke", requestId: 2, invocation: createInvocation("fast-event", "udpDatagram", { bytes: [2], delayMs: 5 }) },
      { type: "stop" },
    ];
    child.stdin.end(messages.map((message) => JSON.stringify(message)).join("\n") + "\n");
    const [exitCode] = await once(child, "exit");
    assert.equal(exitCode, 0);
    assert.deepEqual(responses.slice(1).map((response) => response.requestId), [2, 1]);
    assert.ok(responses.slice(1).every((response) => typeof response.requestId === "number"));
    assert.equal(await readFile(probePath, "utf8"), "stopped");
  } finally {
    await rm(temporaryDirectory, { recursive: true, force: true });
  }
});
